sql: introduce MVCC-compliant index backfiller

stevendanna opened the pull request:
Previously, the index backfilling process depended upon non-MVCC
compliant AddSSTable calls which potentially rewrote previously read
historical values.

To support an MVCC-compliant AddSSTable that writes at the _current_
timestamp, this change implements a new backfilling process described
in the following RFC:

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20211004_incremental_index_backfiller.md

In summary, the new index backfilling process depends on backfilling
the new index when it is in a BACKFILLING state (added in #72281). In
this state it receives no writes or deletes. Writes that occur during
the backfilling process are captured by a "temporary index."  This
temporary index uses the DeletePreservingEncoding to ensure it
captures deletes as well as writes.

After the of bulk backfill using the MVCC-compliant AddSSTable, the
index is moved into a MERGING state
(added in #75663) in which it receives writes and deletes. Writes
previously captured by the temporary index are then transactionally
merged into the newly added index.

This feature is currently behind a new boolean cluster setting which
default to true. Schema changes that contains both old and new-style
backfills are rejected.

Reverting the default to false will require updating various tests
since many tests depend on the exact index IDs of newly added indexes.

Release note: None

stevendanna committed:
sql: introduce MVCC-compliant index backfiller

Previously, the index backfilling process depended upon non-MVCC
compliant AddSSTable calls which potentially rewrote previously read
historical values.

To support an MVCC-compliant AddSSTable that writes at the _current_
timestamp, this change implements a new backfilling process described
in the following RFC:

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20211004_incremental_index_backfiller.md

In summary, the new index backfilling process depends on backfilling
the new index when it is in a BACKFILLING state (added in #72281). In
this state it receives no writes or deletes. Writes that occur during
the backfilling process are captured by a "temporary index."  This
temporary index uses the DeletePreservingEncoding to ensure it
captures deletes as well as writes.

After the of bulk backfill using the MVCC-compliant AddSSTable, the
index is moved into a MERGING state
(added in #75663) in which it receives writes and deletes. Writes
previously captured by the temporary index are then transactionally
merged into the newly added index.

This feature is currently behind a new boolean cluster setting which
default to true. Schema changes that contains both old and new-style
backfills are rejected.

Reverting the default to false will require updating various tests
since many tests depend on the exact index IDs of newly added indexes.

Release note: None

Co-authored-by: Rui Hu <rui@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73878)
<!-- Reviewable:end -->

Commit:
sql: distribute the index merging process

This distributes and checkpoints the index merging process. The
merging process checkpoint is per temporary index.

Release note: None

Co-authored-by: Steven Danna <danna@cockroachlabs.com>

stevendanna commented:
@dt There are still some TODOs here, but I was wondering if you want to give this a review pass?

dt commented on pkg/sql/schema_changer.go:
```diff
@@ -1095,9 +1189,14 @@ func (sc *SchemaChanger) done(ctx context.Context) error {
-				if err := sc.createIndexGCJob(ctx, idx.GetID(), txn, description); err != nil {
-					return err
+				if isTemporaryIndex(idx) {
+					if err := sc.createTemporaryIndexGCJob(ctx, idx.GetID(), txn, "TEMPORARY INDEXES USED DURING INDEX BACKFILL"); err != nil {
```
Is the all caps to style? fine with it if it is, just checking

stevendanna replied:
Thanks for catching this, I meant to fix it up, I made it all caps to make it easier to spot in assertion failures in some of the logic tests and then didn't change it back.

dt reviewed (commented):
I did a once through but defer to a schema reviewer on most of this that is pretty much pure schema code. My only real question is about if we actually care about initialReadTimestamp

stevendanna commented on pkg/sql/mvcc_backfiller.go:
```diff
@@ -70,6 +72,10 @@ func (sc *SchemaChanger) Merge(
 
 			wb := txn.NewBatch()
 			for i := range kvs {
+				if kvs[i].Value.Timestamp.Less(readAsOf) {
```
I believe it is required for correctness. I've explained the case it is handling in the commit message, but can also put a comment here.

dt replied:
Ah. Okay, caught up on the commit message and now see why it's here. Let me ponder a bit, since I'm not sure this will work on top of restored data where we forward all timestamps.

stevendanna replied:
Yeah.  Another option (echoing from slack) is to persist the timestamp in the body of a preserved delete so that it could survive the restore.

stevendanna replied:
This has been made unnecessary by not issuing delete-preserving deletes when the index is in DeleteOnly. See #75720

stevendanna commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -647,7 +647,12 @@ message SchemaChangeDetails {
+  // InitialReadTimestamp is the first read timestamp of the bulk
+  // index backfill. Writes in the temporary index below this
+  // timestamp are ignored by the index merging process.
+  util.hlc.Timestamp initial_read_timestamp = 11 [(gogoproto.nullable) = false];
```
Yup, that is the only user of this.

stevendanna replied:
This has been made unnecessary by not issuing delete-preserving deletes when the index is in DeleteOnly. See https://github.com/cockroachdb/cockroach/pull/75720

stevendanna commented on pkg/sql/schema_changer.go:
```diff
@@ -2520,6 +2661,24 @@ func (sc *SchemaChanger) getDependentMutationsJobs(
+		return false
+	}
+
+	if m.Adding() && idx.IsSharded() && !isTemporaryIndex(idx) {
```
No real reason, this was a recent addition to the schema changer so I should probably follow up with schema here.

ajwerner commented on pkg/sql/backfill.go:
```diff
@@ -1842,19 +1850,93 @@ func ValidateForwardIndexes(
+// checked to ensure all newly added indexes are using one type of
+// index backfill.
+//
+// The MVCC-copatible index backfilling process has a goal of not
```
nit: copatible

ajwerner commented on pkg/sql/catalog/tabledesc/index.go:
```diff
@@ -346,6 +346,16 @@ func (w index) ForcePut() bool {
+// TODO(ssd): This could be its own boolean or we could store the ID
+// of the index it is a temporary index for.
+func (w index) IsTemporaryIndexForBackfill() bool {
+	return w.desc.UseDeletePreservingEncoding
```
➕ it might be nice to be able to associate the two indexes.

ajwerner commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -2069,11 +2131,51 @@ func (desc *Mutable) AddComputedColumnSwapMutation(swap *descpb.ComputedColumnSw
+		m.State = descpb.DescriptorMutation_DELETE_AND_WRITE_ONLY
+	}
+	desc.addMutationWithNextID(m)
+	// If we are adding an index, we add another mutation for the
+	// temporary index used by the index backfiller.
+	//
+	// The index backfiller code currently assumes that it can
+	// always find the temporary indexes in the Mutations array,
+	// in same order as the adding indexes.
+	if idxMut, ok := m.Descriptor_.(*descpb.DescriptorMutation_Index); ok {
```
does this need version gating?

ajwerner commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -74,6 +74,15 @@ message BackfillerSpec {
 
   optional int32 initial_splits = 11 [(gogoproto.nullable) = false];
+
+    // WriteAtRequestTimestamp controls the corresponding AddSSTable request
```
nit: indentation

ajwerner commented on pkg/sql/mvcc_backfiller.go:
let's file an issue for this?

ajwerner commented on pkg/sql/schema_changer.go:
```diff
@@ -987,10 +1047,102 @@ func (sc *SchemaChanger) RunStateMachineBeforeBackfill(ctx context.Context) erro
+				continue
+			}
+
+			if m.Adding() {
+				if m.Backfilling() {
+					tbl.Mutations[m.MutationOrdinal()].State = descpb.DescriptorMutation_DELETE_ONLY
+					runStatus = RunningStatusDeleteOnly
+				} else if m.DeleteOnly() {
+					tbl.Mutations[m.MutationOrdinal()].State = descpb.DescriptorMutation_MERGING
+					runStatus = RunningStatusMerging
+				}
+			}
+		}
+		if runStatus == "" || tbl.Dropped() {
+			return nil
+		}
```
this and the stepping twice all feels a little brittle. is there a test which resumes between the two steps?

stevendanna replied:
I've added a test that randomly pauses before calls to `(*schemaChanger).txn` and then resume the job and makes sure it succeeds. It's a little slow so it only does one pause by default, but locally I've run it after every txn:

```
--- PASS: TestPauseBeforeRandomDescTxn (23.87s)                                                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_1 (1.77s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_2 (1.78s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_3 (1.74s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_4 (1.80s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_5 (1.91s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_6 (1.72s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_7 (1.77s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_8 (1.72s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_9 (1.67s)                                                                                                                                                                              
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_10 (1.82s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_11 (1.87s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_12 (1.94s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_13 (1.74s)   
```

I think we need to add more assertions that we don't have any unexpected cruft and it would also be nice to add some concurrent operations, but hopefully this gets at the basic problem you are pointing out here.

ajwerner reviewed (commented):
This PR wasn't very surprising, which is good. On this first pass, I mostly just had nits and a couple of questions. I'm inclined to say we should merge this with some urgency and then keep iterating.

ajwerner commented on pkg/sql/backfill/mvcc_index_merger.go:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	err := ibm.flowCtx.Cfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		// For now just grab all of the destination KVs and merge the corresponding entries.
+		kvs, err := txn.Scan(ctx, key, endKey, chunkSize)
```
If I were being nit-picky, I'd encourage you to use the lower-level APIs. These high-level APIs are sort of garbage. I'd consider something like:

```
var ba roachpb.BatchRequest
	ba.TargetBytes = 16 << 10
	ba.MaxSpanRequestKeys = chunkSize
	ba.Add(&roachpb.ScanRequest{
		RequestHeader: roachpb.RequestHeader{
			Key:    key,
			EndKey: endKey,
		},
		ScanFormat: roachpb.KEY_VALUES,
	})
	br, pErr := txn.Send(ctx, ba)
	if pErr != nil {
		return pErr.GoError()
	}
	resp := br.Responses[0].GetScan()
	for _, row := range resp.Rows {
		
	}
```
Otherwise you're liable to pull a lot more data than you wanted.

stevendanna replied:
Happy to update to this now.  Was 16Kb just a bet guess or do you have some thinking there I can include in a comment?

stevendanna replied:
Decided to push this to a follow up PR. Will open it shortly after this merges.

ajwerner commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -957,23 +957,55 @@ func (h *distSQLNodeHealth) check(ctx context.Context, sqlInstanceID base.SQLIns
 	partitions := make([]SpanPartition, 0, 1)
+	var userDataByPartition [][]interface{}
+
+	getUserData := func(i int) (interface{}, error) {
+		if userData == nil {
+			return nil, nil
+		}
+
+		if i < 0 || i >= len(userData) {
+			return nil, errors.Errorf("cannot get userData with index %d", i)
+		}
+		return userData[i], nil
+	}
```
it feels like you may as well just check that userData and spans are the same size and then just use the same `i` value in the loop below. This feels contrived.

stevendanna replied:
Your other suggestion removes this code completely, so nothing more to do here.

ajwerner commented on pkg/sql/distsql_plan_backfill.go:
```diff
@@ -83,3 +100,55 @@ func (dsp *DistSQLPlanner) createBackfillerPhysicalPlan(
+) (*PhysicalPlan, error) {
+
+	var indexSpans []roachpb.Span
+	var spanIdxs []interface{}
+
+	for i := range spans {
+		for j := range spans[i] {
+			indexSpans = append(indexSpans, spans[i][j])
+			spanIdxs = append(spanIdxs, i)
+		}
+	}
+	spanPartitions, idxByPartition, err := dsp.PartitionSpansWithUserData(planCtx, indexSpans, spanIdxs)
```
this all feels gross. I really don't like `PartitionSpansWithUserData`. Probably with as much code, and fewer allocations than you've infected this code with, you could use a data structure to look up the index:

```go
type spanAndIndex struct {
    roachpb.Span
    idx int
}

var _ interval.Interface = (*spanAndIndex)(nil)

func (si *spanAndIndex) Range() inteval.Range { return si.AsRange() }
func (si *spanAndIndex) ID() uintptr { return uintptr(unsafe.Pointer(si)) }

func (dsp *DistSQLPlanner) createIndexBackfillerMergePhysicalPlan(
	planCtx *PlanningCtx, spec execinfrapb.IndexBackfillMergerSpec, spans [][]roachpb.Span,
) (*PhysicalPlan, error) {
        var n int
	for _, sp := range spans {
		for range sp {
		        n++
		}
	}
	indexSpans := make([]roachpb.Span, 0, n)
        spanIdxs := make([]spanAndIdxs, 0, n)
        spanIdxTree := interval.NewTree(interval.ExclusiveOverlapper)
        for i := range spans {
		for j := range spans[i] {
			indexSpans = append(indexSpans, spans[i][j])
			spanIdxs = append(spanIdxs, spanAndIndex{Span: spans[i][j], idx: i})
			spanIdxTree.Insert(&spanIdxs[len(spanIdxs)-1], true /* fast */)
		}
	}
	spanIdxTree.AdjustRanges()
	spanPartitions, err := dsp.PartitionSpansWithUserData(planCtx, indexSpans)
	// ...
	// now you can look up an index for a span like:
	getIndex := func(sp roachpb.Span) (idx int) {
	    if !spanIdxTree.DoMatching(func(i Interface) (done bool) {
	        i.(*spanAndIndex).idx
	        return false
	     }, sp.AsRange()) {
	        panic("didn't find it")
	     }
	     return idx
	}
	
```

stevendanna replied:
Thanks for this suggestion.  I've updated the PR to use this method. @rhu713 Mind taking a quick look at the last few commits to make sure you are OK with those changes.

ajwerner reviewed (approved):
Sorry it took me so long to actually make my way through this. My comments are largely superficial and can be made later as cleanup. Let's get this merged so we can gain experience with it

stevendanna commented:
bors r=ajwerner
